### PR TITLE
Avoid exception when converting from pandas adjacency with integer column names

### LINKED ISSUE #7407: networkx.exception.NetworkXUnfeasible when creating a graph from pandas
### Current Behavior

When creating a graph from a pandas dataframe, if the row/columns names are integer less than the dimensionality of the frame, an error occurs:
```
  raise nx.NetworkXUnfeasible(
networkx.exception.NetworkXUnfeasible: The node label sets are overlapping and no ordering can resolve the mapping. Use copy=True.
```
### Expected Behavior
Graph is created without errors
To fix it:
`nx.relabel.relabel_nodes(G, dict(enumerate(df.columns)), copy=False`) in convert_matrix.py should have copy=True otherwise there's a duplication of labels

### Steps to Reproduce

```
M = pd.DataFrame(0.0, index=[1010001,2,1,1010002], columns=[1010001,2,1,1010002])
G = nx.from_pandas_adjacency(M, create_using = nx.DiGraph)
```

Possible fix appears in #7408 

### Environment

Python version: 3.9, 3.10
NetworkX version: 3.1, 3.3

### PR DESCRIPTION
**Pull request by katerinakazantseva**
Fixes #7407

### CODE REVIEW

**Comment by dschult** [comment]
Thanks for this -- nice catch and thanks for tracking down the bug! 

Unfortunately, switching `copy` to `True` in the `relabel_nodes` call can be expensive for big graphs and is backward incompatible (as the failing tests show).

What we want here is an option:  `copy=False_if_possible`. Or renaming, something like: `in_place_if_possible=True`. But I don't like either of those names.  But we can build our own in this code rather than changing the function signature of relabel_nodes.

```python
    try:
        return nx.relabel_nodes(G, dict(enumerate(df.columns)), copy=False)
    except nx.NetworkXUnfeasible as err:
        return nx.relabel_nodes(G, dict(enumerate(df.columns)), copy=True)
```
The exception is raised before any relabeling starts, so this is safe from that perspective.
Can you check that this does what you want it to do? And then make the change and push it up to this branch?

Thanks!

**Comment by katerinakazantseva** [comment]
Hi Dan, thanks for the quick response! I just updated the branch, please take a look.

File: networkx/convert_matrix.py [hunk 1]
```diff
@@ -215,8 +215,10 @@ def from_pandas_adjacency(df, create_using=None):
 
     A = df.values
     G = from_numpy_array(A, create_using=create_using)
-
-    nx.relabel.relabel_nodes(G, dict(enumerate(df.columns)), copy=True)
+    try:
+        nx.relabel.relabel_nodes(G, dict(enumerate(df.columns)), copy=False)
+    except nx.NetworkXUnfeasible as err:
+        nx.relabel_nodes(G, dict(enumerate(df.columns)), copy=True)
```
Comment (dschult) [inline networkx/convert_matrix.py:221]: When `copy=True` the relabel_nodes function does not change G. So we need to capture the output of this function call.  We can also capture the output that way even when `copy=False`. So I think we should just return the result in both cases.

Also, let's use `nx.relabel_nodes` in both calls as well.

```python
    try:
        return nx.relabel_nodes(G, dict(enumerate(df.columns)), copy=False)
    except nx.NetworkXUnfeasible as err:
        return nx.relabel_nodes(G, dict(enumerate(df.columns)), copy=True)
```

**Review by dschult — COMMENTED** [review]
Can you add your nice short example from the issue as a test?
It would go in `networkx/tests/test_convert_pandas.py` and can be a function at the bottom rather than a method in the test class (it doesn't use the setup provided by the class. Separate function also makes it essentially a copy/paste of your example wrapped in a function. 

Add a comment in the test like `# check for no exception. See gh-7407`

**Review by rossbar — COMMENTED** [review]
I was thinking about this a bit yesterday, and there may be a really nice solution by adding a feature to `from_numpy_array`.

`from_pandas_adjacency` currently [achieves the result in two steps](https://github.com/networkx/networkx/blob/e97634ab839f7eb40e4887c4b59d13c267977df3/networkx/convert_matrix.py#L217-L219): 1. create the graph from the numpy adjacency matrix, then 2. relabel those nodes. If `from_numpy_array` supported a `nodes` (name tbd) kwarg, then these two steps could be achieved at once and the issue of relabeling inplace would be avoided altogether.

Anyways, just an idea - I haven't actually tried it yet. Don't let it be a blocker for the patches here!

**Comment by dschult** [comment]
I like the idea of `from_numpy_array` supporting a nodelist kind of argument! Especially if it could be done in an integrated way with the graph creation rather than relabeling nodes after-the-fact.

I think this is also a very small step toward @eriknw's interest in forming a standard binary/array-based representation of a graph because connecting the nodes to arrays is a key part of that IMO.

**Comment by dschult** [comment]
This is expanded on via the suggestions @rossbar makes above. 
I will close this in favor of #7412.   

Thanks very much @katerinakazantseva for this improvement!

**Pull request closed without merging**